### DISCUSSION TITLE
Using 'repeat' isn't useful for inserting text

### DISCUSSION THREAD

**Discussion by ideasman42** [meow-edit/meow#414] [General]
In VIM it's common to insert (or change) text, leave insert mode, navigate to a new location in normal mode, then repeat the previous action.

The problem with doing this in meow is the navigation commands (e.g. `meow-prev` or `meow-next`) are repeated.

What is a convenient way to repeat an insertion elsewehre?

**Comment by DogLooksGood** [comment]
For change, you can copy the inserted text after exiting insert mode. But there's no way to duplicate the insertion, unless using kmacro.

**Comment by okamsn** [comment]
I haven't tried it yet, but I have been looking at [the package Dot Mode](https://github.com/wyrickre/dot-mode).

From the README:

> It emulates the vi `redo’ command, repeating the immediately preceding sequence of commands. This is done by recording input commands which change the buffer, i.e. not motion commands.

**Comment by okamsn** [comment]
It's still not as simple as than Evil's `.`, but my experience has been made better by the improvements to the built-in Edmacro Mode in Emacs 30. This is what I am doing now for the complex cases:

- Bind `kmacro-edit-lossage` to `"`.
- Set the new `edmacro-reverse-key-order` to `t` so that newer keys are shown first.
- Select the recent keys I want, then use `edmacro-set-macro-to-region-lines` to delete the keys I don't want.
- Run `edmacro-finish-edit` to create the new keyboard macro.
- Use that new macro like Evil's `.`.

Using Meow's keypad feature, the process is:

- Press `"`.
- Go to a position near the top of the listing and spam `meow-line` (I have it on `o`).
- Press `SPC c r`.
- Press `SPC c c`.
- Use the macro with `F4` like normal.

A wrapper command could combine `SPC c r` and `SPC c c`, but I haven't done that so far. The kmacro approach has the benefit of not being limited to the most recent change.

There's also the new options `meow-select-on-insert` and `meow-select-on-append` for the simple cases.

**Comment by ideasman42** [comment]
Since posting this thread I've developed a package to repeat which can group multiple commands:

- https://github.com/meow-edit/meow/discussions/731
- https://codeberg.org/ideasman42/emacs-repeat-fu
